crypto/bls/blst/signature.go

```diff
@@ -0,0 +1,229 @@
+// Copyright 2023 The klaytn Authors
+// This file is part of the klaytn library.
+//
+// The klaytn library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The klaytn library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the klaytn library. If not, see <http://www.gnu.org/licenses/>.
+
+package blst
+
+import (
+	"crypto/rand"
+	"fmt"
+	"sync"
+
+	"github.com/klaytn/klaytn/crypto/bls/types"
+	"github.com/pkg/errors"
+)
+
+type signature struct {
+	// Pointer to underlying blst struct, hence the name 'p'
+	p *blstSignature
+}
+
+func SignatureFromBytes(b []byte) (types.Signature, error) {
+	if len(b) != types.SignatureLength {
+		return nil, types.ErrSignatureLength(len(b))
+	}
+
+	if s, ok := signatureCache.Get(cacheKey(b)); ok {
+		return s.(*signature).Copy(), nil
+	}
+
+	p := new(blstSignature).Uncompress(b)
+	if p == nil || !p.SigValidate(false) {
+		return nil, types.ErrSignatureUnmarshal
+	}
+
+	s := &signature{p: p}
+	signatureCache.Add(cacheKey(b), s.Copy())
+	return s, nil
+}
+
+func MultipleSignaturesFromBytes(bs [][]byte) ([]types.Signature, error) {
+	if len(bs) == 0 {
```

Comment:
shouldn't this be an error? is this an exception?

Reply:
Within the klaytn/bls/blst package,

- (a) functions converting Multiple bytes -> multiple objects: returns nil if given slice is empty
  - MultiplePublicKeysFromBytes
  - MultipleSignaturesFromBytes
- (a) functions converting Multiple bytes or objects -> aggregated object: obviously can't accept an empty slice
  - AggregatePublicKeys
  - AggregatePublicKeysFromBytes
  - AggregateSignatures
  - AggregateSignaturesFromBytes

It could be an error, but not necessary.